examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb

```diff
@@ -0,0 +1,1907 @@
+            "\u001b[0;31mAttributeError\u001b[0m: 'DeferredDataFrame' object has no attribute 'get_dummies'"
+          ]
+        }
+      ],
+      "source": [
+        "object_class_col= beam_df.filter(items=['object_class'])\n",
+        "object_class_col.get_dummies()"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 321
+        },
+        "id": "hz8s7z8caTq-",
+        "outputId": "5c543a6b-0ea1-41f8-afec-9691bbbd1f5b"
+      },
+      "outputs": [
+        {
+          "output_type": "error",
+          "ename": "AttributeError",
+          "evalue": "ignored",
+          "traceback": [
+            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
+            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
+            "\u001b[0;32m/tmp/ipykernel_325/1927971370.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;31m# df['categories_concat'].str.get_dummies('-')\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m \u001b[0mobject_class_col\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_dummies\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
+            "\u001b[0;32m/content/beam/sdks/python/apache_beam/dataframe/frames.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   2482\u001b[0m       \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2483\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2484\u001b[0;31m       \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2485\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2486\u001b[0m   \u001b[0;32mdef\u001b[0m \u001b[0m__getitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
+            "\u001b[0;31mAttributeError\u001b[0m: 'DeferredDataFrame' object has no attribute 'str'"
+          ]
+        }
+      ],
+      "source": [
+        "object_class_col.str.get_dummies()"
```

Comment:
`str` is an attribute on Series, you need to access just a Series (e.g. `object_class_col = beam_df.object_class`) in order to use `str.get_dummies`. Selecting a single column with `filter` produces a single column DataFrame.

You will also need to configure that column to have a `CategoricalDtype`. This is where you can use `unique` and `ib.collect`, something like:
```
unique_classes = ib.collect(object_class_col.unique())
object_class_col.astype(pd.CategoricalDtype(unique_classes)).str.get_dummies()
```

Reply:
I tried this approach and I am still not able to run the `unique()` command event when defining the argument `unique(as_series=True)` as specified [here](https://beam.apache.org/releases/pydoc/current/apache_beam.dataframe.frames.html#:~:text=unique(as_series%3DFalse,%5Bsource%5D). 
(Error: `unique()` is not implemented for deferred Dataframes`).
In any case, I think it would be nicer to wait up on the `get_dummies()` as it will be more intuitive to use for user. I see that you already filed a [ticket](https://github.com/apache/beam/issues/22646) for it and referred to it in the deliverable 3 doc.

Reply:
Did you try `unique` on just the series that you are trying to one-hot encode? Note that even `pd.get_dummies`, the user is going to have to create a `CategoricalDtype` somehow, which will likely utilize `unique`, so we need to get that working regardless.

Reply:
To be clear, I'm suggesting you change `object_class_col= beam_df.filter(items=['object_class'])` to `object_class_col= beam_df['object_class']`

When you select the column with `filter` it's creating a DataFrame with a single column, not a series.

Reply:
Also it looks like the pandas `DataFrame` doesn't actually have a `unique` operation. If the Beam Deferred DataFrame is giving "unique() is not implemented for deferred Dataframe" then that's a bug. It should give an `AttributeError` like pandas does.

To be sure, I just checked and we do give the same error as pandas:
```
In [7]: df.unique
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [7], in <cell line: 1>()
----> 1 beam_df.unique

File ~/working_dir/beam/sdks/python/apache_beam/dataframe/frames.py:2484, in DeferredDataFrame.__getattr__(self, name)
   2482   return self[name]
   2483 else:
-> 2484   return object.__getattribute__(self, name)

AttributeError: 'DeferredDataFrame' object has no attribute 'unique'
```

If a user wants the unique values in a `DataFrame` they can use `drop_duplicates`, as suggested in https://stackoverflow.com/questions/43184491/df-unique-on-whole-dataframe-based-on-a-column.

`drop_duplicates` (with `keep='any'`) should work in Beam.

Reply:
@TheNeuralBit indeed you're correct `unique()` can only be applied to `Series`. I was able to get the unique classes of the categorical column and transform it to a `CategoricalDtype`:

```
object_class_col= beam_df['object_class']

unique_classes = pd.CategoricalDtype(ib.collect(object_class_col.unique(as_series=True)))
```
This returns 

```
CategoricalDtype(categories=['MBA', 'OMB', 'MCA', 'AMO', 'IMB', 'TJN', 'CEN', 'APO',
                  'ATE', 'AST'], ordered=False)
```

I have tried implementing the workaround  you suggested for `str.get_dummies()` but still ran into some issues:

```
object_class_col.astype(unique_classes).str.get_dummies()

WontImplementError: astype(dtype='category') is not supported because the type of the output column depends on the data. Please use pd.CategoricalDtype with explicit categories instead.
For more information see https://s.apache.org/dataframe-non-deferred-columns.
```
I ran into the same issue when trying to covert the `object_class_col` to a categorical variable:

`
object_class_col.astype('category')
`

Am I still missing something?

Reply:
Hm that's odd. This could be a bug. @yeandy do we have a way to apply a categorical DType to a columnn?

Reply:
Interesting apparently instances of CategoricalDtype are considered equal to the string "category" : https://github.com/pandas-dev/pandas/blob/54347fe684e0f7844bf407b1fb958a5269646825/pandas/core/dtypes/dtypes.py#L366

The aim in our check was to avoid the case where users indicate `astype("category")` and rely on pandas to resolve the categories, since we need explicit categories. We should be able to find another way to check this, but it will have to be another bugfix.

Reply:
[Here](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/dataframe/frames_test.py#L2417) and [here](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/dataframe/frames_test.py#L1486) are some examples of applying `CategoricalDtype`

Reply:
Ah but both of those examples apply the categorical type with astype on a pandas Series, outside of the test framework. It seems we don't verify it with DeferredSeries.astype

Reply:
Filed #23276 to track this bug.

Reply:
The issue with astype should be resolved now and the fix will be in Beam 2.43.0. Can you try this again?

Reply:
thanks @TheNeuralBit ! Indeed this fix seems to the job. I'll submit a PR soon with the latest updates